### PR TITLE
structured attrs: prefer `NIX_ATTRS_*_FILE` over `.attrs.*`

### PR DESCRIPTION
**Pull request by Ma27**
## Description of changes

tl;dr relying on `$NIX_BUILD_TOP/.attrs.*` is problematic because it doesn't exist in `nix-shell`/`nix develop`. While most of `nix-shell` works "by accident" already, this change ensures consistent environments in regular builds and interactive debugging sessions using `nix-shell`/`nix develop`.

For that, references of `.attrs.sh`/`.attrs.json` are replaced (where possible) in a backwards-compatible manner (Nix 2.3 is the minimum Nix supported and doesn't provide these vars) in

* stdenv itself (first commit)
* all drvs using `stdenv` and a custom `builder` (second commit)
* a few drvs using `__structuredAttrs` already and use a custom `builder` (third commit)

See also https://github.com/NixOS/nix/pull/9032.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/stdenv/generic/setup.sh [hunk 1]
```diff
@@ -16,29 +16,15 @@ if (( "${NIX_DEBUG:-0}" >= 6 )); then
...
 
-    # Before Nix 2.4, $NIX_ATTRS_*_FILE was named differently:
-    # https://github.com/NixOS/nix/commit/27ce722
-    if [[ -n "${ATTRS_JSON_FILE:-}" ]]; then
-        export NIX_ATTRS_JSON_FILE="$ATTRS_JSON_FILE"
-    fi
-
-    if [[ -n "${ATTRS_SH_FILE:-}" ]]; then
-        export NIX_ATTRS_SH_FILE="$ATTRS_SH_FILE"
-    fi
```
Comment (Artturin) [inline pkgs/stdenv/generic/setup.sh:40]: Was added in https://github.com/nixos/nixpkgs/commit/1a29857b8a93f5259f0c2e919becc0bf9db24f85

@sternenseemann

Comment (sternenseemann) [reply pkgs/stdenv/generic/setup.sh:40]: > Nix 2.3 is the minimum Nix supported and doesn't provide these vars

This is not quite true, just use `ATTRS_SH_FILE` and `ATTRS_JSON_FILE` as in this removed snippet. Should be kept.

Comment (Ma27) [reply pkgs/stdenv/generic/setup.sh:40]: @sternenseemann first of all sorry, I've seen that you were the author of that change and I intended to request a review from you, but I somehow forgot to add you to the list of reviewers. :facepalm: 

> just use ATTRS_SH_FILE and ATTRS_JSON_FILE as in this removed snippet

I added those in https://github.com/NixOS/nix/commit/3944a120ec6986c723bf36bfade9b331dd4af68a and AFAICS this never reached 2.3. More context is in https://github.com/NixOS/nixpkgs/pull/257919/commits/6884aba7363b1b4f2028a8a69976dd2cf55a5721.

File: pkgs/stdenv/generic/setup.sh [hunk 2]
```diff
@@ -16,29 +16,15 @@ if (( "${NIX_DEBUG:-0}" >= 6 )); then
...
     echo "structuredAttrs is enabled"
-else
-    __structuredAttrs=
-fi
 
-if [ -n "$__structuredAttrs" ]; then
     for outputName in "${!outputs[@]}"; do
         # ex: out=/nix/store/...
         export "$outputName=${outputs[$outputName]}"
     done
```
Comment (lheckemann) [inline pkgs/stdenv/generic/setup.sh:27]: Shouldn't this still be conditional on structuredAttrs? Nix will set the output name environment variables itself, and in fact I don't think we'll have an associative array `outputs` in the non-structuredAttrs scenario?

**Review by lheckemann — APPROVED** [review]

File: pkgs/stdenv/generic/setup.sh (hunk 1 above)
Comment (sternenseemann) [reply pkgs/stdenv/generic/setup.sh:40]: Right, seems like I got indeed fooled with the diff! The workaround code for `NIX_ATTRS_*_FILE` being empty seems to be in place still, so this should be unproblematic for Nix 2.3.

**Comment by Artturin** [comment]
@ofborg build tests.stdenv.structuredAttrsByDefault tests.stdenv.structuredAttrsByDefault.hooks
